### PR TITLE
Alfoa/codeinterfaces

### LINKED ISSUE #446: update Relap5 code interface
--------
Issue Description
--------
Currently, the relap5 code interface will print raven data into csv file, and read it back. This data should be removed.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #1079: Duplicate values in the database
--------
Defect Description
--------

**Describe the defect**

##### What did you expect to see happen?
Being able to sync time series with duplicated valeus 

##### What did you see instead?
Hi guys, I encountered a problem using RAVEN with a Database with duplicate values of the variable "Time".
I can load the database in the RAVEN framework, but I cannot use some post-processor methods such as the HistorySetSync, as well as Externa PP. 
RAVEN returns me the following error:
**ValueError: cannot reindex or align along dimension 'Time' because the index has duplicate values**.

In my specific case, I fixed the problem adding to the MELCOR interface the Pandas dataframe.drop_duplicates() method. However, I think it would be useful to add in RAVEN a post-processor method to remove duplicates from DataBases.

Thanks.

Matteo

##### Do you have a suggested fix for the development team?
N/A

**Describe how to Reproduce**
1. Duplicated values in a time series
2. Apply Sync PP

**Screenshots and Input Files**
Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

**Platform (please complete the following information):**
 - OS: Linux
 - Version: N/A
 - Dependencies Installation: CONDA

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #1274: [DEFECT] RELAP5 Interface - number of digits
--------
Defect Description
--------

**Describe the defect**

##### What did you expect to see happen?
When using the Sampling algorithms, RAVEN should print in the RELAP5-3D input deck a reasonable number of decimals (max 5).

##### What did you see instead?
The interface is printing in the RELAP5-3D input deck the perturbed words with 15 decimals. This will result in a card length > 80 allowed characters, if multiple words are perturbed. This will stop RELAP5-3D execution.

##### Do you have a suggested fix for the development team?
Reduce the number of decimals of the perturbed parameters to <=5. OR Suggest putting a flag/option that can allow the user to select the number of decimals.

**Describe how to Reproduce**
1. 
2. 
3. 
4. 

**Screenshots and Input Files**
Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

**Platform (please complete the following information):**
 - Version: [e.g. 22]
 - Dependencies Installation: [CONDA or PIP]

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #1360: [DEFECT] Datasets unable to handle multiple values at the same time. (e.g. events in Modelica for example)
--------
Defect Description
--------

**Describe the defect**
In the Code Interface for the bouncing ball. The test was created with the values test_Dymola_code_interface_timedep.xml if multiple monte carlo seeds are attempted in the ensemble model. 

##### What did you expect to see happen?
I would expect to be able to collapse a dataset based on the timesteps given out by the code. 

##### What did you see instead?
An error saying the dataset cannot be collapsed along the time variable as 

##### Do you have a suggested fix for the development team?
Not clear.  I think Andrea mentioned the possibility of moderately editing the time variable on a relative basis of 1e-4. (e.g. if there are two steps at time 0.1seconds then the first one would be 0.1 and the second would be 0.10001. Not changing the trajectory, but giving the dataset multiple values to pivot along. 

**Describe how to Reproduce**
1. Have a running version of Modelica on the machine.
2. Unzip the file. 
3. Update the test_Dymola_code_interface_timedep.xml to point to the proper executables. 
4. Try and run.

**Screenshots and Input Files**
Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

[DMInterfaceTestTimedep.zip](https://github.com/idaholab/raven/files/5460626/DMInterfaceTestTimedep.zip)

**Platform (please complete the following information):**
Windows 10.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #1367: [TASK] The codeinterfaces should be able to return the data (outputs) without a CSV
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
The codeinterfaces should be able to return the data (outputs) without a CSV creation (if not wanted)

**Describe the solution you'd like**
finalizeCodeOutput should be able to return a dictionary for the data (in this way we avoid double IO)

**Describe alternatives you've considered**
n/a

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1360 
Closes #1274
Closes #1079 (the duplicated values must be handled at the code interface level)
Closes #446
Closes #1367

##### What are the significant changes in functionality due to this change request?
Addition of the possibility to return the output data directly from the finalizeCodeOutput. In this way,
we avoid the double IO of 
reading Code Output -> memory -> creation of CSV -> RAVEN reading the CSV -> output in memory

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Optimizers/Optimizer.py [hunk 1]
```diff
@@ -400,7 +400,7 @@ def _initializeInitSampler(self, externalSeeding):
                                    .format(v=sampled, i=self._initSampler.name, s=self.name))
     self._initSampler.initialize(externalSeeding)
     # initialize points
-    numTraj = len(self._initialValues) if self._initialValues else None
+    numTraj = len(self._initialValues)
```
Comment (joshua-cogliati-inl) [inline framework/Optimizers/Optimizer.py:403]: Is this part of a different pull request?

File: framework/utils/randomUtils.py [hunk 2]
```diff
@@ -24,10 +24,9 @@
 
 from utils.utils import findCrowModule
 from utils import mathUtils
-import math
 
-# in general, we will use Crow for now, but let's make it easy to switch just in case it is helpful eventually.
-# Numpy stochastic environment can not pass the test as this point
+# in general, we will use Crow for now, but let's make it easy to switch just in case it is helpfull eventually.
+# Numoy stochastic environment can not pass the test as this point
```
Comment (joshua-cogliati-inl) [inline framework/utils/randomUtils.py:29]: The changes on this file revert Diego's random number fixes.

File: scripts/TestHarness/testers/RavenFramework.py [hunk 3]
```diff
@@ -168,7 +168,7 @@ def check_runnable(self):
     ## required module is present, but too old
     if _notQAModules and _checkVersions:
       self.set_fail('skipped (Incorrectly versioned python modules: ' +
-                    " ".join(['{}-{}!={}'.format(*m) for m in _notQAModules]) +
+                    " ".join(['{}-{}'.format(*m) for m in _notQAModules]) +
```
Comment (joshua-cogliati-inl) [inline scripts/TestHarness/testers/RavenFramework.py:171]: This reverts useful changes.  Can you double check that you didn't accidentally do this elsewhere in the pull request?

File: scripts/TestHarness/testers/RavenPython.py [hunk 4]
```diff
@@ -160,7 +160,7 @@ def check_runnable(self):
       return False
     if len(notQa) > 0 and library_handler.checkVersions():
       self.set_fail('skipped (Incorrectly versioned python modules: ' +
-                    " ".join(['{}-{}!={}'.format(*m) for m in notQa]) +
+                    " ".join(['{}-{}'.format(*m) for m in notQa]) +
```
Comment (joshua-cogliati-inl) [inline scripts/TestHarness/testers/RavenPython.py:163]: Reverts useful changes.

File: tests/framework/Optimizers/GradientDescent/init_sampler.xml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" ?>
 <Simulation verbosity="debug">
   <TestInfo>
-    <name>framework/Optimizers/ConvergeGradient.InitSampler</name>
+    <name>framework/Optimizers.ConvergeGradient</name>
```
Comment (joshua-cogliati-inl) [inline tests/framework/Optimizers/GradientDescent/init_sampler.xml:4]: Hm, what is the purpose of this change?

File: tests/framework/pca_LHS/gold/stratifiedCdf/testPrintPointSet_dump.csv [hunk 6]
```diff
@@ -1,10 +1,10 @@
 x0,y0,z0,x1,y1,z1,x,y,z,ProbabilityWeight,PointProbability
```
Comment (joshua-cogliati-inl) [inline tests/framework/pca_LHS/gold/stratifiedCdf/testPrintPointSet_dump.csv:1]: Why are a lot of these files changing?

File: tests/framework/unit_tests/utils/testRandomUtils.py [hunk 7]
```diff
@@ -214,7 +214,7 @@ def checkType(comment,value,expected,updateResults=True):
 ### randomIntegers(), sampling integers in a range
 randomUtils.randomSeed(42,engine=None)
 randomUtils.randomSeed(42,engine=eng)
-right = [14,18,20,12,18]
+right = [14,18,20,12,17]
```
Comment (joshua-cogliati-inl) [inline tests/framework/unit_tests/utils/testRandomUtils.py:217]: I think the changes in this file are a mistake.

**Review by joshua-cogliati-inl — CHANGES_REQUESTED** [review]
There are several things that need changing.  Someone will have to re-review after they are made.

The basic changes look good, but there are several accidental changes that need to be reverted.

File: framework/CodeInterfaceBaseClass.py [hunk 8]
```diff
@@ -46,6 +46,7 @@ def __init__(self):
     self._runOnShell = True      # True if the specified command by the code interfaces will be executed through shell.
     self._ravenWorkingDir = None # location of RAVEN's main working directory
     self._csvLoadUtil = 'pandas' # utility to use to load CSVs
+    self._writeCSV = False       # write CSV even if the data can be returned directly to raven (e.g. if the user requests them)
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/CodeInterfaceBaseClass.py:49]: This is a great idea. I did not think about it. I will add the option in the base class (yes it is not mandatory anymore)

File: framework/CodeInterfaces/RELAP5/RELAPparser.py [hunk 9]
```diff
@@ -144,7 +143,7 @@ def modifyOrAdd(self,dictionaryList,save=True):
         temp.append('*'+' deckNum: '+str(deckNum)+'\n')
       for j in sorted(modiDictionaryList):
         for var in modiDictionaryList[j]:
-          temp.append('* card: '+j+' word: '+str(var['position'])+' value: '+str(var['value'])+'\n')
+          temp.append('* card: '+j+' word: '+str(var['position'])+' value: '+'{:.5e}'.format(var['value'])+'\n')
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/CodeInterfaces/RELAP5/RELAPparser.py:146]: This is just in the comment section of the perturbed relap5 input deck. It is not used in relap. 
It is just to have a better readability of the perturbed file.

File: framework/CodeInterfaces/WorkshopExamples/ProjectileInterface.py [hunk 10]
```diff
@@ -96,6 +97,24 @@ def createNewInput(self, currentInputFiles, oriInputFiles, samplerType, **Kwargs
...
+      @ Out, data, list, the list (timesteps) of list of data [[],[],[],etc]
+    """
+    with open(outfileName, 'r') as src:
+      headers = [x.strip() for x in  src.readline().split() ]
+      data = []
+      line = ""
+      # starts reading
+      while not line.strip().startswith("--"):
+        line = src.readline()
+        if not line.strip().startswith("--"):
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/CodeInterfaces/WorkshopExamples/ProjectileInterface.py:114]: much more elegant...fixed

File: framework/CodeInterfaces/WorkshopExamples/ProjectileInterfaceNoCSV.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+class ProjectileNoCSV(Projectile):
+  """
+    Provides code to interface RAVEN to Projectile without the need of a CSV
+    The only method that changes is the finalizeCodeOutput (we return the data directly)
+  """
+  def finalizeCodeOutput(self, command, output, workingDir):
+    """
+      Called by RAVEN to modify output files (if needed) so that they are in a proper form.
+      In this case, the default .mat output needs to be converted to .csv output, which is the
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/CodeInterfaces/WorkshopExamples/ProjectileInterfaceNoCSV.py:40]: fixed. Thanks

File: framework/Models/Code.py [hunk 12]
```diff
@@ -591,23 +592,29 @@ def evaluateSample(self, myInput, samplerType, kwargs):
     ## My guess is that every code interface implements this given that the code
     ## below always adds .csv to the filename and the standard output file does
     ## not have an extension. - (DPM 4/6/2017)
-    outputFile = codeLogFile
+    outputFile, isStr = codeLogFile, True
     if 'finalizeCodeOutput' in dir(self.code) and returnCode == 0:
       finalCodeOutputFile = self.code.finalizeCodeOutput(command, codeLogFile, metaData['subDirectory'])
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/Models/Code.py:597]: right. Done. Thanks

File: framework/Optimizers/Optimizer.py (hunk 1 above)
Comment (alfoa) [reply framework/Optimizers/Optimizer.py:403]: I am sorry. Reverted

File: scripts/TestHarness/testers/RavenFramework.py (hunk 3 above)
Comment (alfoa) [reply scripts/TestHarness/testers/RavenFramework.py:171]: Sorry. Reverted.

File: scripts/TestHarness/testers/RavenPython.py (hunk 4 above)
Comment (alfoa) [reply scripts/TestHarness/testers/RavenPython.py:163]: Sorry. Reverted.

File: tests/framework/Optimizers/GradientDescent/init_sampler.xml (hunk 5 above)
Comment (alfoa) [reply tests/framework/Optimizers/GradientDescent/init_sampler.xml:4]: Sorry a mistake

File: framework/Models/Code.py [hunk 13]
```diff
@@ -591,23 +592,29 @@ def evaluateSample(self, myInput, samplerType, kwargs):
...
+        isStr = isinstance(finalCodeOutputFile,string_types)
+        if not isDict and not isStr:
+          self.raiseAnError(RuntimeError, 'The return argument from "finalizeCodeOutput" must be either a str' +
+                                          'containing the new output file root or a dict of data!')
       if finalCodeOutputFile and not ravenCase:
-        outputFile = finalCodeOutputFile
+        if not isDict:
+          outputFile = finalCodeOutputFile
+        else:
+          returnDict = finalCodeOutputFile
```
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/Models/Code.py:611]: Done

File: framework/utils/randomUtils.py (hunk 2 above)
Comment (alfoa) [reply framework/utils/randomUtils.py:29]: reverted....sorry for all of these

File: tests/framework/pca_LHS/gold/stratifiedCdf/testPrintPointSet_dump.csv (hunk 6 above)
Comment (alfoa) [reply tests/framework/pca_LHS/gold/stratifiedCdf/testPrintPointSet_dump.csv]: reverted the files

File: tests/framework/unit_tests/utils/testRandomUtils.py (hunk 7 above)
Comment (alfoa) [reply tests/framework/unit_tests/utils/testRandomUtils.py]: fixed

File: framework/CodeInterfaceBaseClass.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/CodeInterfaceBaseClass.py:49]: Added this feature. Thank you for the advice

**Comment by alfoa** [comment]
@joshua-cogliati-inl  I addressed your comments.

**Comment by joshua-cogliati-inl** [comment]
> @joshua-cogliati-inl I addressed your comments.

Thanks, I'll take a look today.

File: framework/utils/randomUtils.py [hunk 14]
```diff
@@ -11,22 +11,22 @@
...
 import numpy as np
+import math
 from collections import deque, defaultdict
 
 from utils.utils import findCrowModule
 from utils import mathUtils
-import math
 
-# in general, we will use Crow for now, but let's make it easy to switch just in case it is helpful eventually.
+# in general, we will use Crow for now, but let's make it easy to switch just in case it is helpfull eventually.
```
Comment (joshua-cogliati-inl) [inline framework/utils/randomUtils.py:29]: For what it is worth, helpful is the more standard spelling.

File: tests/framework/CodeInterfaceTests/PHISICS/gold/phisicsRelap5/phisicsRelap5Test/1/ex_3a_R_CR_nom.i [hunk 15]
```diff
@@ -41,7 +41,7 @@
 2500000  inlet  tmdpvol
 2500101  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0000000
 2500200  003
-2500201  0.0  6.39e+06  530.245239193
+2500201  0.0  6.39e+06  5.30245e+02
```
Comment (joshua-cogliati-inl) [inline tests/framework/CodeInterfaceTests/PHISICS/gold/phisicsRelap5/phisicsRelap5Test/1/ex_3a_R_CR_nom.i:44]: Is this a comment or a piece of data that is used by Relap5?

**Review by joshua-cogliati-inl — APPROVED** [review]
This is much improved.  I approve.

**Comment by joshua-cogliati-inl** [comment]
When the tests pass, this is ready to go.
